### ISSUE TITLE
  Wrap more operators in CoalesceBatchExec

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#139]
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-11068

Once [https://github.com/apache/arrow/pull/9043] is merged, we should extend this to wrap HashJoinExec and HashAggregateExec as well since they can both produce small batches.

Rather than hard-code a list of operators that need to be wrapped, we should find a more generic mechanism so that plans can declare if their input and/or output batches should be coalesced (similar to how we handle partitioning) and this would allow custom operators outside of DataFusion to benefit from this optimization.

**Comment by alamb** [comment]
Comment from Andrew Lamb(alamb) @ 2020-12-30T11:37:27.904+0000:
<pre>I have some suggestions here: https://github.com/apache/arrow/pull/9043#discussion_r550163707

TLDR -- I think incorporating the coalescing logic into the operators themselves (so they don't produce small baches in the first place) might be better</pre>

**Comment by alamb** [comment]
Has had no activity for several years, let's close it down and reopen if it is ever still relevant

I think all operators that would benefit from CoalesceBatches have been wrapped now
